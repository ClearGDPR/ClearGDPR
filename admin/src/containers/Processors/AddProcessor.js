import React from 'react';
import PropTypes from 'prop-types';

import config from 'config';
import internalFetch from '../../helpers/internal-fetch';
import { PanelConsumer } from '../MainLayout/PanelContext';
import EditProcessorForm from '../../components/Processors/EditProcessor';

export class AddProcessorContainer extends React.Component {
  static propTypes = {
    processor: PropTypes.object,
    closePanel: PropTypes.func
  };

  state = {
    isLoading: false,
    errors: {}
  };

  async updateProcessor(processor) {
    await internalFetch(`${config.API_URL}/api/management/processors/add`, {
      method: 'POST',
      body: JSON.stringify(processor)
    }).catch(err => {
      this.setState({
        errors: {
          '': err.message
        }
      });
      return Promise.reject(err);
    });
  }

  startLoading() {
    this.setState({
      isLoading: true
    });
  }

  stopLoading() {
    this.setState({
      isLoading: false
    });
  }

  onSubmit() {
    this.startLoading();

    return this.updateProcessor(this.props.processor)
      .then(this.stopLoading.bind(this))
      .then(() => this.props.closePanel && this.props.closePanel())
      .catch(this.stopLoading.bind(this));
  }

  render() {
    return (
      <EditProcessorForm
        errors={this.state.errors}
        onSubmit={this.onSubmit.bind(this)}
        isLoading={this.state.isLoading}
      />
    );
  }
}

export default props => (
  <PanelConsumer>
    {({ closePanel }) => <AddProcessorContainer {...props} closePanel={closePanel} />}
  </PanelConsumer>
);
