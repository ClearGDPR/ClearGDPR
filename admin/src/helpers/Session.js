import cookie from 'cookie';

export class Session {
  set(authResult) {
    const maxAge = 1 * 24 * 60 * 60; // 1 day long;
    document.cookie = cookie.serialize('auth', authResult, { maxAge });
    document.cookie = cookie.serialize('expires_at', authResult, { maxAge });
  }

  destroy() {
    document.cookie = cookie.serialize('auth', '', { maxAge: -1 });
    document.cookie = cookie.serialize('expires_at', '', { maxAge: -1 });
  }

  isLoggedIn() {
    let expiresAt = cookie.parse(document.cookie).expires_at || '';
    return new Date().getTime() < expiresAt;
  }
}

export default new Session();
