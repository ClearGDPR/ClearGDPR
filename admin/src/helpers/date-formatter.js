import { format } from 'date-fns';

export default date => {
  return format(new Date(date), 'DD/MM/YYYY h:mma');
};
