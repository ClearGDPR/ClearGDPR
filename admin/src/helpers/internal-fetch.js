import merge from 'lodash/object/merge';
import session from '../../helpers/Session';

const internalFetch = (url, config = {}) => {
  const internalConfig = merge(
    {
      method: 'GET',
      headers: {
        Authorization: `Bearer ${session.getToken()}`,
        'Content-Type': 'application/json'
      }
    },
    config
  );
  return fetch(url, internalConfig).then(async res => {
    if (res.status === 400 && (await res.json()).error === 'JWT token expired') {
      session.destroy();
    } else {
      return res;
    }
  });
};

export default internalFetch;
