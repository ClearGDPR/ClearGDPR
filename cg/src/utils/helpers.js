const Promise = require('bluebird');
const winston = require('winston');

const { VALID_RUN_MODES } = require('./constants');

function inControllerMode() {
  return process.env.MODE === VALID_RUN_MODES.CONTROLLER;
}

function getMyAddress() {
  return process.env.MY_ADDRESS;
}
function getTokenExpiry() {
  return process.env.JWT_TOKEN_EXPIRY || 604800;
}

const timeout = ms => new Promise(res => setTimeout(res, ms));

async function retryAsync(promise, options = {}) {
  let tries = 0;

  const maxRetries = options.maxRetries || 3;
  const interval = options.interval || 500;
  let delay = 0;

  while (tries < maxRetries) {
    await Promise.delay(delay);
    tries++;
    try {
      return await promise();
    } catch (e) {
      delay = interval * tries;
      winston.info(`Attempt no. ${tries} failed. Next attempt in ${delay}ms`);
      winston.log('debug', `Error: ${e.toString()}`);
    }
  }

  winston.error(`Operation failed after ${tries} attempts`);
}

module.exports = {
  getTokenExpiry,
  inControllerMode,
  getMyAddress,
  timeout,
  retryAsync
};
