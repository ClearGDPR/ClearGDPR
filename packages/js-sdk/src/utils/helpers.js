import { merge } from 'lodash/object';

export function parseJSON(response) {
  const contentType = response.headers.get('Content-Type');

  if (contentType.indexOf('application/json') > -1) {
    return new Promise(resolve =>
      response.json().then(json => {
        return resolve({
          status: response.status,
          ok: response.ok,
          data: json
        });
      })
    );
  } else {
    return new Promise(resolve =>
      response.text().then(text => {
        return resolve({
          status: response.status,
          ok: response.ok,
          data: text
        });
      })
    );
  }
}

export function buildRequestBody(body) {
  return body ? JSON.stringify(body) : null;
}

/**
 * Merges objects values
 * @param {Object} options objects to be merged
 * @returns {Object} New merged object based on params
 */
export function mergeOptions(...objects) {
  return merge(...objects);
}
