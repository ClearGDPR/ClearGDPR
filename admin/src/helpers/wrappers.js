import React from 'react';
import PropTypes from 'prop-types';
import lodash from 'lodash';
import { UsersProvider } from 'containers/Users/UsersContext';
import { ProcessorsProvider } from 'containers/Processors/ProcessorsContext';
import { PanelProvider } from 'containers/MainLayout/PanelContext';
import { RectificationsProvider } from 'containers/Rectifications/RectificationsContext';
import { StatsProvider } from 'containers/Dashboard/StatsContext';

export const withWrapper = Wrapper => Component => {
  let wrapper = props => (
    <Wrapper>
      <Component {...props}>{props.children}</Component>
    </Wrapper>
  );
  wrapper.propTypes = {
    children: PropTypes.node
  };
  return wrapper;
};

export const withContextProviders = lodash.flow.apply(null, [
  withWrapper(UsersProvider),
  withWrapper(ProcessorsProvider),
  withWrapper(RectificationsProvider),
  withWrapper(PanelProvider),
  withWrapper(StatsProvider)
]);
