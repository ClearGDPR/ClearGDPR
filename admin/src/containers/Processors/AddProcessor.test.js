import React from 'react';
import { shallow } from 'enzyme';

import * as TestUtils from 'tests/helpers/TestUtils';

import { AddProcessorContainer } from './AddProcessor';

const setupShallow = propOverrides => {
  const props = Object.assign(
    {
      addProcessor: jest.fn().mockReturnValue(Promise.resolve()),
      closePanel: jest.fn(),
      isLoading: false
    },
    propOverrides
  );
  const component = shallow(<AddProcessorContainer {...props} />);

  return { props, component };
};

describe('(Container) Add Processor', () => {
  it('should render correctly when default props provided', async () => {
    const { component } = setupShallow();
    expect(component).toMatchSnapshot();
  });

  it('should populate errors on submit when create rejects with error', async () => {
    const { props, component } = setupShallow();
    const stubProcessor = {
      name: null,
      logoUrl: null,
      scopes: []
    };

    props.addProcessor.mockReturnValue(Promise.reject(new Error('Test error')));
    component.instance().onSubmit(stubProcessor);

    await TestUtils.flushPromises();
    expect(component.state()).toMatchSnapshot();
  });

  it('should close panel when create succeeded', async () => {
    const { props, component } = setupShallow();
    const stubProcessor = {
      name: 'Processor Name',
      logoUrl: 'https://a_valid_url.com/file.png',
      scopes: []
    };

    component.instance().onSubmit(stubProcessor);

    await TestUtils.flushPromises();
    expect(props.closePanel).toHaveBeenCalled();
  });

  it('should be able to add processor without scopes', () => {
    const { component } = setupShallow();
    const stubProcessor = {
      name: 'Processor Name',
      logoUrl: 'https://a_valid_url.com/file.png'
    };
    try {
      component.instance().onSubmit(stubProcessor);
    } catch (e) {
      expect(true).toBe(false);
    }
  });
});
