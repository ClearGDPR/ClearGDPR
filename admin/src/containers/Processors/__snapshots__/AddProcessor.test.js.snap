// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`(Container) Add Processor should populate errors on submit when create processor rejects with error 1`] = `
Object {
  "errors": Object {
    "username": "Error: Test error",
  },
}
`;

exports[`(Container) Add Processor should render correctly when default props provided 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <AddProcessorContainer
    addProcessor={[MockFunction]}
    closePanel={[MockFunction]}
    isLoading={false}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "errors": Object {},
      "isLoading": undefined,
      "onSubmit": [Function],
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "errors": Object {},
        "isLoading": undefined,
        "onSubmit": [Function],
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
